c10/core/impl/cow/shadow_storage.h

```diff
@@ -0,0 +1,70 @@
+#pragma once
+
+// This file provides two public types: ShadowStorage and
+// ShadowStorageNonIntrusive.
+//
+// See ShadowStorageImpl to understand the interface of these types.
+
+#include <c10/macros/Macros.h>
+#include <c10/util/intrusive_ptr.h>
+
+#include <atomic>
+#include <cstdint>
+#include <type_traits>
+
+namespace c10::impl::cow::detail {
+
+// An empty struct to use as the base class for a non-intrusive
+// ShadowStorageImpl.
+struct Empty {};
+
+// Represents the shadow storage for a tensor.
+//
+// For simulated lazy copies, the shadow storage represents the what
+// the storage would have been for a tensor that was lazily
+// copied. However, when simulating this, we can't actually replace
+// the storage because we would violate a core implementation
+// invariant that tensors which are views of each other share the same
+// storage.
+//
+// However, the fidelity of this check is limited by the extent to
+// which we have instrumented operations as reading or writing to
+// storage.
+//
+// We use a monotonically increasing generation number to track
+// modifications to storage.
+//
+// Note that we templatize on whether or not we are eligible to be
+// allocated into an intrusive_ptr. It *seems* as though ASan is
+// unhappy if we stack-allocate an object that inherits from
+// c10::intrusive_ptr_target.
```

Comment:
What exactly is the ASAN error?  See

```
// Note [Stack allocated intrusive_ptr_target safety]
// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// A well known problem with std::enable_shared_from_this is that it
// allows you to create a std::shared_ptr from a stack allocated object,
// which is totally bogus because the object will die once you return
// from the stack.  In intrusive_ptr, we can detect that this has occurred,
// because we set the refcount/weakcount of objects which inherit from
// intrusive_ptr_target to zero, *unless* we can prove that the object
// was dynamically allocated (e.g., via make_intrusive).
//
// Thus, whenever you transmute a T* into a intrusive_ptr<T>, we check
// and make sure that the refcount isn't zero (or, a more subtle
// test for weak_intrusive_ptr<T>, for which the refcount may validly
// be zero, but the weak refcount better not be zero), because that
// tells us if the object was allocated by us.  If it wasn't, no
// intrusive_ptr for you!
```

Reply:
i removed the templating on intrusiveness so we'll see if it is still happening.

Reply:
This did not recur on the latest manifestation. I'm going to assume that I had some other programming error that caused this that no longer exists.